File: day01/main.py
from pathlib import Path

here = Path(__file__).absolute().parent


def get_total_calories(s: str) -> list[int]:

    elves = s.strip().split("\n\n")
    total_calories = [sum(map(int, elf.split("\n"))) for elf in elves]
    total_calories.sort()
    return total_calories


def main() -> int:

    with open(here / "input.txt") as f:
        text = f.read()

    total_calories = get_total_calories(text)
    print("Part 1:")
    print(total_calories[-1])
    print("Part 2:")
    print(sum(total_calories[-3:]))

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
else:
    import pytest

    @pytest.mark.parametrize(
        ("input_s", "expected"),
        (("""1\n2\n\n4\n""", [3, 4]), ("""10\n2\n\n4\n""", [4, 12])),
    )
    def test_get_total_calories(input_s: str, expected: int) -> None:
        assert get_total_calories(input_s) == expected

File: day01/src/main.rs
use std::{collections::BinaryHeap, fs};

fn str2int(cal: &str) -> u32 {
    cal.parse::<u32>().unwrap()
}

fn elf2total_calories(elf: &str) -> u32 {
    elf.split("\n").map(str2int).sum()
}

fn get_sorted_calories(s: String) -> BinaryHeap<u32> {
    let elves = s.trim_end().split("\n\n");
    let total_calories = elves.map(elf2total_calories);
    let sorted_calories: BinaryHeap<u32> = total_calories.collect();
    sorted_calories
}

fn main() {
    let file_path = "input.txt";
    println!("In file {}", file_path);

    let contents = fs::read_to_string(file_path).expect("Should have been able to read the file");
    let mut sorted_calories = get_sorted_calories(contents);

    let mut total = sorted_calories.pop().unwrap();
    println!("Part 1:\n{total}");

    for _n in 1..3 {
        total += sorted_calories.pop().unwrap();
    }
    println!("Part 2:\n{total}");
}
